JEP358 extended NPE message part 1 - result of this NPE

JasonFengJ9 opened the pull request:
**JEP358 extended NPE message part 1 - result of this NPE**

- `VM_GetExtendedNPEMessage()` is updated to invoke `iterateStackTrace()` with `getStackTraceElementIterator()` to get the `romClass/romMethod/bytecodeOffset` of first frame of the exception object then invoke `getCompleteNPEMessage()` for message generation;
- Added `getCompleteNPEMessage()` in a new file `vm/ExtendedMessageNPE.cpp` to compute `NPE` message according to the
`bytecode` itself for `aload/astore, arraylength, athrow, monitorenter/monitorexit`, the bytecode and field index for `getfield/putfield`, the `bytecode` and method index for `invokeinterface/invokespecial/invokevirtual`;
- Added a runtime flag `J9_EXTENDED_RUNTIME2_SHOW_EXTENDED_NPEMSG` to specify the command line option `-XX:{+|-}ShowCodeDetailsInExceptionMessages`;
Added trace messages.

Sample NPE messages (`-Xint` or `-Xjit:count=0`):
```
java.lang.NullPointerException: Cannot load from int array
	at a.se.java_lang_npe.NullPointerExceptionTest.testFailedAction(NullPointerExceptionTest.java:143)

java.lang.NullPointerException: Cannot store to object array
	at a.se.java_lang_npe.NullPointerExceptionTest.testFailedAction(NullPointerExceptionTest.java:254)

java.lang.NullPointerException: Cannot read the array length
	at a.se.java_lang_npe.NullPointerExceptionTest.testFailedAction(NullPointerExceptionTest.java:325)

java.lang.NullPointerException: Cannot enter synchronized block
	at a.se.java_lang_npe.NullPointerExceptionTest.testFailedAction(NullPointerExceptionTest.java:344)

java.lang.NullPointerException: Cannot read field "nullInstanceField"
	at a.se.java_lang_npe.NullPointerExceptionTest.testFailedAction(NullPointerExceptionTest.java:357)

Output:
java.lang.NullPointerException: Cannot assign field "i"
	at a.se.java_lang_npe.NullPointerExceptionTest.test_aload(NullPointerExceptionTest.java:622)

java.lang.NullPointerException: Cannot invoke "a/se/java_lang_npe/G.m2(Ljava/lang/String;)Ljava/lang/String;"
	at a.se.java_lang_npe.NullPointerExceptionTest.testFailedAction(NullPointerExceptionTest.java:412)

java.lang.NullPointerException: Cannot invoke "a/se/java_lang_npe/NullPointerExceptionTest$A.getB()La/se/java_lang_npe/NullPointerExceptionTest$B;"
	at a.se.java_lang_npe.NullPointerExceptionTest.testMixedChasing(NullPointerExceptionTest.java:860)

```

related: #8450 
Issue: #7500 

Reviewer: @DanHeidinga 
FYI: @gacholio 

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

gacholio commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -9621,6 +9623,13 @@ runMethodHandle: {
+nullPointerWithExtendedMsg:
+	updateVMStruct(REGISTER_ARGS);
+	prepareForExceptionThrow(_currentThread);
+	setCurrentExceptionUTF(_currentThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, getExtendedNPEMessage(_currentThread, *_pc, *(U_16*)(_pc + 1), J9_CP_FROM_METHOD(_literals)));
```
Might be better to simply pass `_currentThread` to the message helper and let it fetch the data that's required.

Also need to verify that every path here has not incremented the pc (I suspect this is already true).

JasonFengJ9 replied:
`prepareForExceptionThrow(_currentThread)` could update `_currentThread->_pc`, hence `getExtendedNPEMessage()` can't retrieve current `_pc` value via `_currentThread`.

JasonFengJ9 replied:
@gacholio @DanHeidinga this is ready for review.

JasonFengJ9 commented:
ci compilation failure
```
/home/travis/build/eclipse/openj9-openjdk-jdk11/openj9/runtime/gc_include/ObjectAccessBarrierAPI.hpp:30:10: fatal error: omrmodroncore.h: No such file or directory
 #include "omrmodroncore.h"
          ^~~~~~~~~~~~~~~~~
compilation terminated.
make[6]: *** [runtime/jit_vm/CMakeFiles/j9jit_vm.dir/cthelpers.cpp.o] Error 1
```

This doesn't appear related to this PR.

JasonFengJ9 commented on runtime/j9vm/j7vmi.c:
```diff
@@ -1358,13 +1358,11 @@ jint JNICALL
-	J9JavaVM * vm = currentThread->javaVM;
-	J9InternalVMFunctions * vmfns = vm->internalVMFunctions;
-	jint numberOfFrames;
-	UDATA pruneConstructors = 0;
```
Got it, restored and made it `const`.

JasonFengJ9 commented on runtime/j9vm/j9scar.tdf:
```diff
@@ -338,3 +338,8 @@ TraceEvent=Trc_SC_LoadSystemLibrary_LoadFailed NoEnv Overhead=1 Level=1 Template
+TraceEntry=Trc_SC_GetExtendedNPEMessage_Entry Overhead=1 Level=3 Template="JVM_GetExtendedNPEMessage for throwableObj(%p)"
+TraceEntry=Trc_SC_GetExtendedNPEMessage_Entry2 Overhead=1 Level=1 Template="JVM_GetExtendedNPEMessage for throwableObj(%p)"
+TraceEvent=Trc_SC_GetExtendedNPEMessage_NULLMETHODPC Overhead=1 Level=1 Template="JVM_GetExtendedNPEMessage iterateStackTrace returns a NULL methodPC."
+TraceExit=Trc_SC_GetExtendedNPEMessage_Exit Overhead=1 Level=3 Template="JVM_GetExtendedNPEMessage returns msgObjectRef(%p)"
```
Yeah, fixed.

JasonFengJ9 commented on runtime/j9vm/j9scar.tdf:
```diff
@@ -338,3 +338,8 @@ TraceEvent=Trc_SC_LoadSystemLibrary_LoadFailed NoEnv Overhead=1 Level=1 Template
+
+TraceEntry=Trc_SC_GetExtendedNPEMessage_Entry Overhead=1 Level=3 Template="JVM_GetExtendedNPEMessage for throwableObj(%p)"
+TraceEntry=Trc_SC_GetExtendedNPEMessage_Entry2 Overhead=1 Level=1 Template="JVM_GetExtendedNPEMessage for throwableObj(%p)"
+TraceEvent=Trc_SC_GetExtendedNPEMessage_NULLMETHODPC Overhead=1 Level=1 Template="JVM_GetExtendedNPEMessage iterateStackTrace returns a NULL methodPC."
```
Fixed.

JasonFengJ9 commented on runtime/nls/j9vm/j9vm.nls:
```diff
@@ -1953,3 +1953,94 @@ J9NLS_VM_CLASS_RELATIONSHIP_INVALID.explanation=The specified child class fails
 J9NLS_VM_CLASS_RELATIONSHIP_INVALID.system_action=The JVM will throw a java/lang/VerifyError.
 J9NLS_VM_CLASS_RELATIONSHIP_INVALID.user_response=Ensure that all class relationships are valid.
 # END NON-TRANSLATABLE
+
```
Sure, removed these NLS messages and put them directly in `exceptionsupport.c`.

JasonFengJ9 commented on runtime/j9vm/j9scar.tdf:
```diff
@@ -338,3 +338,8 @@ TraceEvent=Trc_SC_LoadSystemLibrary_LoadFailed NoEnv Overhead=1 Level=1 Template
 TraceException=Trc_SC_RegisterSignal_FailedToRegisterHandler Overhead=1 Level=1 Template="Failed to register handler: OS signal value = %d, new signal handler = %p, old signal handler = %p"
+
+TraceEntry=Trc_SC_GetExtendedNPEMessage_Entry Overhead=1 Level=3 Template="JVM_GetExtendedNPEMessage for throwableObj(%p)"
+TraceEntry=Trc_SC_GetExtendedNPEMessage_Entry2 Overhead=1 Level=1 Template="JVM_GetExtendedNPEMessage for throwableObj(%p)"
```
`Trc_SC_GetExtendedNPEMessage_Entry`  is invoked for each `JVM_GetExtendedNPEMessage()`,
`Trc_SC_GetExtendedNPEMessage_Entry2` is invoked only when `-XX:+ShowCodeDetailsInExceptionMessages` is specified.
The thought was that `Trc_SC_GetExtendedNPEMessage_Entry` is the higher frequency tracepoint than `Trc_SC_GetExtendedNPEMessage_Entry2`.  The hot tracepoint is assigned with higher level, i.e., `3` while the lower frequency one is `1`.

JasonFengJ9 commented on runtime/oti/vm_api.h:
```diff
@@ -554,10 +554,11 @@ internalExceptionDescribe(J9VMThread *vmThread);
 */
 UDATA
-iterateStackTrace(J9VMThread * vmThread, j9object_t* exception,  UDATA  (*callback) (J9VMThread * vmThread, void * userData, J9ROMClass * romClass, J9ROMMethod * romMethod, J9UTF8 * fileName, UDATA lineNumber, J9ClassLoader* classLoader), void * userData, UDATA pruneConstructors);
+iterateStackTrace(J9VMThread * vmThread, j9object_t* exception,  UDATA  (*callback) (J9VMThread * vmThread, void * userData, U_8 *methodPC, J9ConstantPool *constantPool, J9ROMClass * romClass, J9ROMMethod * romMethod, J9UTF8 * fileName, UDATA lineNumber, J9ClassLoader* classLoader), void * userData, UDATA pruneConstructors, U_32 nbrCacheItems);
```
Fixed.

@DanHeidinga please have another look.

JasonFengJ9 replied:
Just realized that `iterateStackTrace()` can't depend on the incoming `numCacheItems` to find the cached `constantPool`, otherwise there is a possibility that some `iterateStackTrace()` calls might not correctly iterate the `walkback` when `NPE` extended message is enabled with additional J9_STACKWALK_CACHE_CPS.

Added a helper method `isNPEMsgEnabled()` to determine if the `constantPool` entry presents or not.

Please review.

JasonFengJ9 commented:
Rebased and resolved the conflict.

JasonFengJ9 commented:
As per @DanHeidinga desk comments that retrieving `methodPC` & `constantPool` from the top stack doesn't warrant the `iterateStackTrace` interface change which affects each caller for all stack frames, the PR is updated to introduce a `iterateStackTraceHelper` that only call back with `methodPC` & `constantPool` for the top frame.
Passed `Java 8/11/14` personal builds.

pshipton commented:
Chatting with Dan I don't expect this is going the make the 0.20.0 release, so I'll move it out.

JasonFengJ9 commented:
**Updates**
1. Get first methodPC of the `exception` `walkback` object without going through `iterateStackTrace` (with some code duplication);
2. Move the message part 1 computing into the verifier which lays the framework for the message part 2 PR;

Tested in a personal build (`JDK 8/11/14` `xa64`/`wa64`).
This PR matches the part 1 message of JTReg test `NullPointerExceptionTest.java` [1].

@DanHeidinga @gacholio please have another look.

[1] https://github.com/ibmruntimes/openj9-openjdk-jdk14/blob/openj9/test/hotspot/jtreg/runtime/exceptionMsgs/NullPointerException/NullPointerExceptionTest.java

gacholio commented:
I think this is far too much duplication of code - it would be much better to tie in to `iterateStackTrace`. The callback allows you to terminate the iteration, so you can operate on just the top frame.

JasonFengJ9 commented:
> I think this is far too much duplication of code - it would be much better to tie in to iterateStackTrace. The callback allows you to terminate the iteration, so you can operate on just the top frame.

Earlier version of this PR https://github.com/eclipse/openj9/commit/f09497bd21932d79192fda0e71e59cba18c858d7 modified `iterateStackTrace` to return `methodPC` via a `getStackTraceElementIterator` however this requires the modification of `callback` interface of `iterateStackTrace` hence its all callers require change to add an extra parameter.

To avoid the modification of all callers of `iterateStackTrace`, also tried https://github.com/eclipse/openj9/commit/a019c474a5188629460c8e02b62afc3acf99c16a which introduced a helper method `iterateStackTraceHelper` and `callback_func_t_once_only`.

To get the `methodPC` via  `iterateStackTrace` callback `getStackTraceElementIterator`, I think one of change above is needed.
Is there another way to avoid the code duplication?

gacholio commented:
I'd prefer you add the parameter than duplicate the code (there aren't all that many users, and the duplicated code is fairly involved).

gacholio commented:
And i assume `methodPC` is actually `bytecodePCOffset` and not an actual pointer (if so, please name it that way in the callback).

JasonFengJ9 commented:
> I'd rather you add the parameter than duplicate the code (there aren't all that many users, and the duplicated code is fairly involved).

Added another parameter `UDATA bytecodeOffset` at `iterateStackTrace` `callback` function, updated other callers accordingly.

Please review.

gacholio commented on runtime/bcverify/bcverify.c:
```diff
@@ -2404,15 +2397,20 @@ j9bcv_verifyBytecodes (J9PortLibrary * portLib, J9Class * clazz, J9ROMClass * ro
 	romMethod = (J9ROMMethod *) J9ROMCLASS_ROMMETHODS(romClass);
 
-	if (verboseVerification) {
+	if (verboseVerification && !isNPEMsg) {
```
Do not use `!` on non-booleans - do the compare properly please, or create a boolean local.

gacholio replied:
Nevermind, I can't read.

gacholio commented on runtime/bcverify/rtverify.c:
```diff
@@ -401,17 +400,187 @@ matchStack(J9BytecodeVerificationData * verifyData, J9BranchTargetStack *liveSta
+/**
+* Return an extended NPE message.
+*
+* Note: the caller is responsible for freeing the returning string if it is not NULL.
```
returning -> returned

JasonFengJ9 replied:
fixed.

gacholio commented on runtime/bcverify/rtverify.c:
```diff
@@ -401,17 +400,187 @@ matchStack(J9BytecodeVerificationData * verifyData, J9BranchTargetStack *liveSta
+* @param bcCurrentPtr The pointer to the bytecode being executed and caused the NPE
+* @param romClass The romClass of the bytecode
+* @param constantPool The J9RomClass constant pool for the romClass
+* @param npeCauseMsg The extended NPE message part 2 (to be provided by another PR)
```
The "part 2" is confusing, and please don't mention "another PR".

JasonFengJ9 replied:
fixed.

gacholio commented on runtime/bcverify/rtverify.c:
```diff
@@ -401,17 +400,187 @@ matchStack(J9BytecodeVerificationData * verifyData, J9BranchTargetStack *liveSta
-
-	returns BCV_SUCCESS on success
-	returns BCV_ERR_INTERNAL_ERROR on verification error
-	returns BCV_ERR_INSUFFICIENT_MEMORY on OOM
```
Looks like you've lost the comment for verifyBytecodes.

gacholio replied:
Nevermind, I see it in the header now.

gacholio commented on runtime/bcverify/rtverify.c:
```diff
@@ -526,6 +714,11 @@ verifyBytecodes (J9BytecodeVerificationData * verifyData)
 
 	/* walk the bytecodes linearly */
 	while (pc < length) {
+		if (isNPEMsg && pc >= npePC) {
```
Why >= ? Should this not be == ?

JasonFengJ9 replied:
There are cases that the `pc` might be increased beyond `npePC` such as the case of `JBinvokeinterface2` right before the bytecode at `npePC`. https://github.com/eclipse/openj9/blob/9edff1b76ed55c0e40efb801dc83bc0a5b7c0a24/runtime/bcverify/rtverify.c#L623 increases current `pc` by `5` hence more than `npePC`.

gacholio commented on runtime/j9vm/javanextvmi.c:
```diff
@@ -32,10 +36,87 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
+	J9ROMClass *romClass;
+	J9ROMMethod *romMethod;
+	UDATA bytecodeOffset;
+} GetStackTraceElementUserData;
```
Blank line please.

JasonFengJ9 replied:
added.

gacholio commented on runtime/j9vm/javanextvmi.c:
```diff
@@ -32,10 +36,87 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
+	userData->bytecodeOffset = bytecodeOffset;
+
+	return FALSE;
+}
```
Blank line please.

JasonFengJ9 replied:
Added.

gacholio commented on runtime/j9vm/javanextvmi.c:
```diff
@@ -32,10 +36,87 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
+	if (J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_SHOW_EXTENDED_NPEMSG)) {
+		J9InternalVMFunctions const * const vmFuncs = vm->internalVMFunctions;
+		char *npeMsg = NULL;
+		GetStackTraceElementUserData userData = {0}; /* only first frame is needed */
```
Comment seems misplaced.

JasonFengJ9 replied:
Removed. (there is already such a comment in `getStackTraceElementIterator`)

gacholio commented on runtime/j9vm/javanextvmi.c:
```diff
@@ -32,10 +36,87 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
+		userData.bytecodeOffset = UDATA_MAX;
+		vmFuncs->iterateStackTrace(vmThread, (j9object_t*)throwableObj, getStackTraceElementIterator, &userData, FALSE);
+		if ((NULL != userData.romClass) && (NULL != userData.romMethod) && (UDATA_MAX != userData.bytecodeOffset)) {
+			J9BytecodeVerificationData *bcvd = vm->bytecodeVerificationData;
```
This needs to be null-checked.

JasonFengJ9 replied:
Fixed.

gacholio commented on runtime/j9vm/javanextvmi.c:
```diff
@@ -32,10 +36,87 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
+				j9object_t msgObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(vmThread, (U_8 *)npeMsg, strlen(npeMsg), 0);
+				if (NULL != msgObject) {
+					msgObjectRef = vmFuncs->j9jni_createLocalRef(env, msgObject);
+					if (NULL == msgObjectRef) {
```
This should move a bit further out - if the internal string allocation fails, you should also set this. Suggest moving to after the memory free.

JasonFengJ9 replied:
Agreed.

gacholio commented on runtime/j9vm/javanextvmi.c:
```diff
@@ -32,10 +36,87 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
+
+	return FALSE;
+}
+static void cfdumpBytecodePrintFunction(void *userData, char *format, ...)
```
What is the purpose of this? It's controlled by a tracepoint, which is very odd.

If this was just debug, I suggest you remove it as it's going to link in unnecessary code.

JasonFengJ9 replied:
It is used by the trace point to dump the bytecode. Removing it won't compile unless the trace point is removed as well.
I can decorate the trace point along with this function with a debug define if the extra code is a concern.

gacholio replied:
Please eliminate the tracepoint - this isn't the way to do debug.

Also please do wrap this in a debug define (I would just output the stuff always if debug is defined, no need for tracepoint).

JasonFengJ9 replied:
Sure, removed the tracepoint and wrapped it with `DEBUG_BCV` (reuse the one in `bcverify.c`).

gacholio commented:
The message code seems not to be using the NLS messages.

JasonFengJ9 commented:
> The message code seems not to be using the NLS messages.

The message code is not using NLS messages as per an earlier comment https://github.com/eclipse/openj9/pull/8495#discussion_r380869385

gacholio commented:
If the messages aren't being used, why are they being defined?

gacholio commented on runtime/bcverify/rtverify.c:
```diff
@@ -526,6 +714,11 @@ verifyBytecodes (J9BytecodeVerificationData * verifyData)
 
 	/* walk the bytecodes linearly */
 	while (pc < length) {
+		if (isNPEMsg && pc >= npePC) {
```
Please bracket the >=

JasonFengJ9 replied:
Added.

JasonFengJ9 commented:
> If the messages aren't being used, why are they being defined?

Aha, that's some leftover from a stale commit, removed.

Please have another look.

gacholio commented on runtime/j9vm/javanextvmi.c:
```diff
@@ -32,10 +40,96 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
+			&& (UDATA_MAX != userData.bytecodeOffset)
+			&& (NULL != bcvd)
+		) {
+#ifdef DEBUG_BCV
```
Please use `#if defined` rather than `#ifdef`, and please comment the else and endif directives (for all ifdefs in this file and any others you may have added).

JasonFengJ9 replied:
Fixed.

gacholio commented:
jenkins test sanity plinux jdk11

gacholio commented:
I'll wait for the OK from @DanHeidinga before merging (or he can do it when he approves).

gacholio commented:
jenkins test extended xlinux jdk14

gacholio commented:
@JasonFengJ9 Merge conflicts have been introduced and will need to be resolved.

JasonFengJ9 commented:
> Merge conflicts have been introduced and will need to be resolved.

Merge conflicts are resolved.

gacholio commented:
jenkins test sanity,extended zlinux jdk14

DanHeidinga commented:
I've been discussing this (and phase 2) with @JasonFengJ9 via Slack.  I really really don't want this code in the verifier as it doesn't belong there.  Verification is already a beast and any extra code in there makes the whole thing more complex to reason about.

For Phase 2, Jason's going to add a new bytecode walker anyway as the info he needs to track (PCs rather than types) is different than what the verifier is set up to provide.  I'd prefer to rebase phase 1 on that bytecode walker rather than merge code into the verifier now.

JasonFengJ9 commented:
@DanHeidinga this PR has been updated to move away from the `verifier`, please review.

JasonFengJ9 commented:
`travis-ci cmake` failure: `CMakeFiles/j9vm.dir/intfunc.c.o:(.data.rel+0x478): undefined reference to getCompleteNPEMessage`.

Added `ExtendedMessageNPE.cpp` to `vm/CMakeLists.txt`.

DanHeidinga commented on runtime/vm/ExtendedMessageNPE.cpp:
```diff
@@ -0,0 +1,201 @@
+			Trc_VM_GetCompleteNPEMessage_Unreachable(vmThread, bcCurrent);
+		}
+		if ((bcCurrent >= JBiaload) && (bcCurrent <= JBsaload)) {
+#define VM_NPE_ALOAD "Cannot load from %s array"
+			nlsMessage = VM_NPE_ALOAD;
+#undef VM_NPE_ALOAD
```
I'm not clear why this is using defines rather than the string directly?
```suggestion
			nlsMessage = "Cannot load from %s array";
```

JasonFengJ9 replied:
It was from an earlier commit in which there were multiple references to the string.
Fixed.

DanHeidinga commented on runtime/vm/ExtendedMessageNPE.cpp:
```diff
@@ -0,0 +1,201 @@
+			nlsMessage = VM_NPE_ALOAD;
+#undef VM_NPE_ALOAD
+		} else {
+#define VM_NPE_ASTORE "Cannot store to %s array"
+			nlsMessage = VM_NPE_ASTORE;
+#undef VM_NPE_ASTORE
```
```suggestion
			nlsMessage = "Cannot store to %s array";
```

JasonFengJ9 replied:
Fixed.

DanHeidinga commented on runtime/vm/ExtendedMessageNPE.cpp:
```diff
@@ -0,0 +1,201 @@
+		j9str_printf(PORTLIB, npeMsg, msgLen, nlsMessage, elementType);
+	} else {
+		switch (bcCurrent) {
+		case JBarraylength:
+#define VM_NPE_ARRAYLENGTH "Cannot read the array length"
+			msgLen = j9str_printf(PORTLIB, NULL, 0, VM_NPE_ARRAYLENGTH);
+			npeMsg = (char *)j9mem_allocate_memory(msgLen + 1, OMRMEM_CATEGORY_VM);
+			j9str_printf(PORTLIB, npeMsg, msgLen, VM_NPE_ARRAYLENGTH);
+#undef VM_NPE_ARRAYLENGTH
```
I'm unclear on why a define is preferable to local variable here.  The local seems to have less "noise" when reading the code
```suggestion
		case JBarraylength:
			const char *VM_NPE_ARRAYLENGTH = "Cannot read the array length";
			msgLen = j9str_printf(PORTLIB, NULL, 0, VM_NPE_ARRAYLENGTH);
			npeMsg = (char *)j9mem_allocate_memory(msgLen + 1, OMRMEM_CATEGORY_VM);
			j9str_printf(PORTLIB, npeMsg, msgLen, VM_NPE_ARRAYLENGTH);
```

JasonFengJ9 replied:
This comes from the thought (and becomes a habit later) in early days that the defines save the memory defining local variables. It is negligible and probably covered by compiler optimization already.
Agreed, the local reads better.

DanHeidinga replied:
Make the local const and it shouldn't matter

DanHeidinga commented on runtime/vm/ExtendedMessageNPE.cpp:
```diff
@@ -0,0 +1,201 @@
+				romMethodRef = (J9ROMMethodRef *)&constantPool[index];
+				nameAndSig = J9ROMMETHODREF_NAMEANDSIGNATURE(romMethodRef);
+				sig = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSig);
+				name = J9ROMNAMEANDSIGNATURE_NAME(nameAndSig);
```
The locals can be declared at first use as this is a cpp file.  Saves initializing them to null and shortens the life time they need to be reasoned about

JasonFengJ9 replied:
Yeah, should use this advantage, updated.

DanHeidinga commented on runtime/vm/exceptiondescribe.c:
```diff
@@ -319,7 +319,7 @@ iterateStackTrace(J9VMThread * vmThread, j9object_t* exception, callback_func_t
 					}
 					if (inlineDepth == 0) {
```
Can you set `methodPC = UDATA_MAX;` here or before this if statement so the else clauses aren't needed?

JasonFengJ9 replied:
There are still a few references of `methodPC` expecting a valid value:
Line 346 
`romClass = findROMClassFromPC(vmThread, methodPC, &classLoader);`, 
Line 370 
`if ((methodPC >= (UDATA)possibleMethod) && (methodPC < (UDATA)J9_BYTECODE_END_FROM_ROM_METHOD(possibleMethod))) {`
Line 381 
`romMethod = findROMMethodInROMClass(vmThread, romClass, methodPC);`

It doesn't seem there is a good place to set `methodPC = UDATA_MAX;` beforehand and save later duplicates.

JasonFengJ9 commented on runtime/j9vm/javanextvmi.c:
```diff
@@ -32,10 +40,93 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
+		Trc_SC_GetExtendedNPEMessage_Entry2(vmThread, throwableObj);
+		vmFuncs->internalEnterVMFromJNI(vmThread);
+		userData.bytecodeOffset = UDATA_MAX;
+		vmFuncs->iterateStackTrace(vmThread, (j9object_t*)throwableObj, getStackTraceElementIterator, &userData, FALSE);
```
@DanHeidinga is it `#define J9_JNI_UNWRAP_REFERENCE(o)  (*(j9object_t*)(o))`? `iterateStackTrace()` expects `j9object_t* exception` though.

JasonFengJ9 replied:
Thanks @DanHeidinga, the PR has been updated, please have another look.

JasonFengJ9 commented on runtime/j9vm/javanextvmi.c:
```diff
@@ -32,10 +40,93 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
+				}
+				j9mem_free_memory(npeMsg);
+				if (NULL == msgObjectRef) {
+					vmFuncs->setNativeOutOfMemoryError(vmThread, 0, 0);
```
Agreed, throwing the `NPE` instead and removed `OOM`.

JasonFengJ9 commented on runtime/vm/ExtendedMessageNPE.cpp:
```diff
@@ -0,0 +1,201 @@
+getCompleteNPEMessage(J9VMThread *vmThread, U_8 *bcCurrentPtr, J9ROMClass *romClass, const char *npeCauseMsg)
+{
+
+	char *npeMsg = NULL;
```
Fixed.

DanHeidinga commented on runtime/oti/vm_api.h:
```diff
@@ -558,7 +558,7 @@ internalExceptionDescribe(J9VMThread *vmThread);
 */
 UDATA
-iterateStackTrace(J9VMThread * vmThread, j9object_t* exception,  UDATA  (*callback) (J9VMThread * vmThread, void * userData, J9ROMClass * romClass, J9ROMMethod * romMethod, J9UTF8 * fileName, UDATA lineNumber, J9ClassLoader* classLoader, J9Class* ramClass), void * userData, UDATA pruneConstructors);
+iterateStackTrace(J9VMThread * vmThread, j9object_t* exception,  UDATA  (*callback) (J9VMThread * vmThread, void * userData, UDATA bytecodeOffset, J9ROMClass * romClass, J9ROMMethod * romMethod, J9UTF8 * fileName, UDATA lineNumber, J9ClassLoader* classLoader, J9Class* ramClass), void * userData, UDATA pruneConstructors);
```
Can you add the new parameter to the `@param` list above?

JasonFengJ9 replied:
Oops, missed this, will add it.

JasonFengJ9 replied:
Renamed `method` to `romMethod`, added `bytecodeOffset`, `romClass`, and `classLoader`.

gacholio commented:
jenkins test sanity plinux jdk11

gacholio commented:
jenkins compile win jdk11

gacholio commented:
jenkins compile zlinux jdk8

gacholio commented:
Previous testing failed for infra reasons and I have no idea how to restart.

gacholio commented:
jenkins test sanity xlinux jdk11

gacholio commented:
I would also say that .cpp files that aren't actually C++ should not being with a capital letter, but that won't hold up this PR.

JasonFengJ9 commented:
The failure was during `git clone` at `pLinux`, and it appears due to https://github.com/eclipse/openj9/issues/9560.

> I would also say that .cpp files that aren't actually C++ should not being with a capital letter

Renamed it to `extendedMessageNPE.cpp`.
P.S. There are quite a few other `.cpp` files with capital letters are not `C++` such as `FastJNI_` friends.

JasonFengJ9 commented:
Earlier PR build links:
jenkins test sanity xlinux jdk11 - https://ci.eclipse.org/openj9/view/Pull%20Requests/job/PullRequest-OpenJ9/3572/
jenkins compile win jdk11 - https://ci.eclipse.org/openj9/view/Pull%20Requests/job/PullRequest-OpenJ9/3567/
jenkins compile zlinux jdk8 - 
https://ci.eclipse.org/openj9/view/Pull%20Requests/job/PullRequest-OpenJ9/3568/

gacholio commented:
> P.S. There are quite a few other `.cpp` files with capital letters are not `C++` such as `FastJNI_` friends.

Yeah, that's my fault - do as I say, not as I do :)

gacholio commented:
I think we've seen enough testing across releases that I won't launch any more (and I verified the cmake file was updated with the new filename).
